core/tx.go

```diff
@@ -260,3 +271,79 @@ func (db *DB) FetchMPWithID(id uint64, accountMP *StateMerkleProof) (err error)
 	*accountMP = accMP
 	return nil
 }
+
+// Validate creates proofs for validating txs and returns new root post validation
+func (tx *Tx) Validate(bz Bazooka, currentRoot ByteArray) (newRoot ByteArray, err error) {
+	fromStateProof, toStateProof, txDBConn, err := tx.GetVerificationData()
+	if err != nil {
+		return
+	}
+
+	newRoot, err = bz.ProcessTx(currentRoot, *tx, fromStateProof, toStateProof)
+	if err != nil {
+		if txDBConn.Instance != nil {
+			txDBConn.Instance.Rollback()
+			txDBConn.Close()
+		}
+		return
+	}
+	if txDBConn.Instance != nil {
+		txDBConn.Instance.Commit()
```

Comment:
What insert or update is committed here?

Reply:
There are updates performed in GetTxVerificationData() via a MySQL transaction, we can rollback those updates if processTx is invalid.